python/tests/test_ownership_variable.py

```diff
@@ -0,0 +1,543 @@
+                                                  unit='s'))
+    assert sc.identical(v_methdeepcopy,
+                        make_variable(1.0, variance=10.0, unit='m'))
+
+
+def test_own_var_scalar_pyobj_set():
+    # Input data is deep-copied.
+    x = {'num': 1, 'list': [2, 3]}
+    s = make_variable(x)
+    x['num'] = -1
+    s.value['list'][0] = -2
+    assert sc.identical(s, make_variable({'num': 1, 'list': [-2, 3]}))
+    assert x == {'num': -1, 'list': [2, 3]}
+
+
+def test_own_var_scalar_pyobj_get():
+    # .value getter shares ownership of the object.
+    s = make_variable({'num': 1, 'list': [2, 3]})
+    x = s.value
+    x['num'] = -1
+    s.value['list'][0] = -2
+    expected = {'num': -1, 'list': [-2, 3]}
+    assert sc.identical(s, make_variable(expected))
+    assert x == expected
+
+
+def test_own_var_scalar_pyobj_copy():
+    # Depth of copies of variables can be controlled.
+    data = {'num': 1, 'list': [2, 3]}
+    s = make_variable(data)
+    s_copy = copy(s)
+    s_deepcopy = deepcopy(s)
+    s_methcopy = s.copy(deep=False)
+    s_methdeepcopy = s.copy(deep=True)
+
+    s.value['num'] = -1
+    s.value['list'][0] = -2
+    modified = {'num': -1, 'list': [-2, 3]}
+    assert sc.identical(s, make_variable(modified))
+    assert sc.identical(s_copy, make_variable(modified))
+    assert sc.identical(s_deepcopy, make_variable(data))
+    assert sc.identical(s_methcopy, make_variable(modified))
+    assert sc.identical(s_methdeepcopy, make_variable(data))
+
+
+def test_own_var_scalar_str_get():
+    # Python strings are immutable, no references are shared.
+    s = make_variable('abc')
+    x = s.value
+    s.value = 'def'
+    assert sc.identical(s, make_variable('def'))
+    assert x == 'abc'
+
+
+def test_own_var_scalar_str_copy():
+    # Depth of copies of variables can be controlled.
+    s = make_variable('abc')
+    s_copy = copy(s)
+    s_deepcopy = deepcopy(s)
+    s_methcopy = s.copy(deep=False)
+    s_methdeepcopy = s.copy(deep=True)
+
+    s.value = 'def'
+    assert sc.identical(s, make_variable('def'))
+    assert sc.identical(s_copy, make_variable('def'))
+    assert sc.identical(s_deepcopy, make_variable('abc'))
+    assert sc.identical(s_methcopy, make_variable('def'))
+    assert sc.identical(s_methdeepcopy, make_variable('abc'))
+
+
+def test_own_var_scalar_nested_set():
+    # Variables are shared when nested.
+    a = np.arange(5)
+    inner = make_variable(a, unit='m')
+    outer = make_variable(inner)
+    outer.value['x', 0] = -1
+    outer.value.values[1] = -2
+    inner['x', 2] = -3
+    inner.unit = 's'
+    a[3] = -4
+    assert sc.identical(
+        outer, make_variable(make_variable([-1, -2, -3, 3, 4], unit='s')))
+    assert sc.identical(inner, make_variable([-1, -2, -3, 3, 4], unit='s'))
+    np.testing.assert_array_equal(a, [0, 1, 2, -4, 4])
+
+    # Assignment creates a new inner Variable.
+    outer.value = make_variable(np.arange(5, 8), unit='kg')
```

Comment:
Is this really new, or rather references the RHS?

Reply:
You are right. Changed the test to check that.